component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -199,4 +209,117 @@ private List<String> populateMandatoryClaims() {
+            String registeredSigningAlgorithm = getRegisteredSigningAlgorithm(getClientId(request, bodyParameters,
+                    oAuthClientAuthnContext));
+            if (registeredSigningAlgorithm.equals(Constants.NOT_APPLICABLE)) {
+                return false;
+            }
+            if (!(Constants.ALG_ES256.equals(requestSigningAlgorithm) ||
+                    Constants.ALG_PS256.equals(requestSigningAlgorithm))) {
+                if (log.isDebugEnabled()) {
+                    log.debug("FAPI unsupported signing algorithm " + requestSigningAlgorithm + " is used to " +
+                            "sign the JWT");
+                }
+                return false;
+            }
+            if (!(isNotEmpty(requestSigningAlgorithm) &&
+                    requestSigningAlgorithm.equals(registeredSigningAlgorithm))) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Registered algorithm does not match with the token signed algorithm");
+                }
+                return false;
+            } else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Obtain the request signing algorithms registered for the application.
+     *
+     * @param clientId   Client ID of the application.
+     * @return Registered signing algorithm for the application.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRegisteredSigningAlgorithm(String clientId) throws OAuthClientAuthnException {
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (!(isNotEmpty(serviceProvider.getCertificateContent()))) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (Constants.TOKEN_ENDPOINT_AUTH_SIGNING_ALG.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return Constants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Obtain the request signing algorithm from the signed JWT.
+     *
+     * @param signedObject   Signed JWT in the request.
+     * @return Request signing algorithm.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRequestSigningAlgorithm(String signedObject) throws OAuthClientAuthnException {
+        //retrieve algorithm from the signed JWT
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(signedObject);
+            return signedJWT.getHeader().getAlgorithm().getName();
+        } catch (ParseException e) {
```

Comment:
IF we r skipping e clause for a reason,shall we least log it?

Reply:
Fixed in c18c85cc171d19c53412ac6eaad0e1e4b738d909